### ISSUE TITLE
Does AffineChannel ops have fp16 support?

### ISSUE THREAD

**Issue by yougoforward** [facebookresearch/Detectron#450]
I'm trying to test detectron faster-rcnn model using  fp16 instead of fp32(np.float32).  I have convert the input data and pretrained params into fp16.
I got error when testing:

`terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at tensor.h:511] IsType<T>(). Tensor type mismatch, caller expects elements to be float while tensor contains caffe2::__f16 Error from operator: 
input: "gpu_0/conv1" input: "gpu_0/res_conv1_bn_s" input: "gpu_0/res_conv1_bn_b" output: "gpu_0/conv1" name: "" type: "AffineChannel" device_option { device_type: 1 cuda_gpu_id: 0 } debug_info: "  File \"/home/long/anaconda/lib/python2.7/threading.py\", line 774, in __bootstrap\n    self.__bootstrap_inner()\n  File \"/home/long/anaconda/lib/python2.7/threading.py\", line 801, in __bootstrap_inner\n    self.run()\n  File \"/home/long/anaconda/lib/python2.7/threading.py\", line 754, in run\n    self.__target(*self.__args, **self.__kwargs)\n  File \"/home/long/objectdetection/fisherDemo.py\", line 199, in camera\n    model = infer_engine.initialize_model_from_cfg()\n  File \"/home/long/github/detectron/lib/core/test_engine.py\", line 275, in initialize_model_from_cfg\n    model = model_builder.create(cfg.MODEL.TYPE, train=False, gpu_id=gpu_id)\n  File \"/home/long/github/detectron/lib/modeling/model_builder.py\", line 124, in create\n    return get_func(model_type_func)(model)\n  File \"/home/long/github/detectron/lib/modeling/model_builder.py\", line 89, in generalized_rcnn\n    freeze_conv_body=cfg.TRAIN.FREEZE_CONV_BODY\n  File \"/home/long/github/detectron/lib/modeling/model_builder.py\", line 229, in build_generic_detection_model\n    optim.build_data_parallel_model(model, _single_gpu_build_func)\n  File \"/home/long/github/detectron/lib/modeling/optimizer.py\", line 54, in build_data_parallel_model\n    single_gpu_build_func(model)\n  File \"/home/long/github/detectron/lib/modeling/model_builder.py\", line 169, in _single_gpu_build_func\n    blob_conv, dim_conv, spatial_scale_conv = add_conv_body_func(model)\n  File \"/home/long/github/detectron/lib/modeling/FPN.py\", line 55, in add_fpn_ResNet50_conv5_body\n    model, ResNet.add_ResNet50_conv5_body, fpn_level_info_ResNet50_conv5\n  File \"/home/long/github/detectron/lib/modeling/FPN.py\", line 112, in add_fpn_onto_conv_body\n    conv_body_func(model)\n  File \"/home/long/github/detectron/lib/modeling/ResNet.py\", line 38, in add_ResNet50_conv5_body\n    return add_ResNet_convX_body(model, (3, 4, 6, 3))\n  File \"/home/long/github/detectron/lib/modeling/ResNet.py\", line 96, in add_ResNet_convX_body\n    p = model.AffineChannel(p, \'res_conv1_bn\', dim=64, inplace=True)\n  File \"/home/long/github/detectron/lib/modeling/detector.py\", line 119, in AffineChannel\n    blob_out = self.net.AffineChannel([blob_in, scale, bias], blob_in)\n  File \"/home/long/github/caffe2/build/caffe2/python/core.py\", line 2047, in <lambda>\n    op_type, *args, **kwargs)\n  File \"/home/long/github/caffe2/build/caffe2/python/core.py\", line 2024, in _CreateAndAddToSelf\n    op = CreateOperator(op_type, inputs, outputs, **kwargs)\n"
** while accessing input: gpu_0/res_conv1_bn_b`

In my experiments, I follow the guide from [https://github.com/caffe2/caffe2/blob/master/caffe2/python/examples/resnet50_trainer.py](url)

**Comment by ir413** [comment]
Hi @yougoforward, current `AffineChannel` op is [instantiated for `float`](https://github.com/caffe2/caffe2/blob/master/modules/detectron/affine_channel_op.cu#L54) so I suspect that you would need to do the same for `float16` or make it polymorphic.

**Comment by JingyuQian** [comment]
@yougoforward May I ask how you managed to convert model and data into fp16 format?

**Comment by yougoforward** [comment]
@JingyuQian 
You can load the .pkl and convert the parameters from fp32 to fp16 like (params).astype(numpy.float16) since the loaded params are just numpy data. Same operations to image data.
If you want to run the detection model as fp16, several ops may not support the fp16, like affinechannel and RoIFeatureTransform(in detector.py), ResizeNearest(in function add_topdown_lateral_module fpn.py),Sigmoid and GenerateProposals(in function add_fpn_rpn_outputs fpn.py). You can use the model.net.HalfToFloat and model.net.FloatToHalf to change the data type when needed as a net layer.

I found the model.net.HalfToFloat and model.net.FloatToHalf cost much time since the number Affinechannel op is the same as the conv op in backbone network. A choice is replacing the affinechannel op with SpatialBN op( BN support fp16) and change the params correspondingly(affinechannel params _s,_b; SpatialBN params _s,_b,_rm=1,_riv=0 ).

Another choice is to fuse affinechannel params with the adjacent conv op, convert conv op without bias to conv with bias.

**Comment by JingyuQian** [comment]
@yougoforward  So you're saying, where these operators occur, I need to add precision conversion before after these operators, right? (I want to try a simple version first, without considering the time cost of HalfToFloat/FloatToHalf)

**Comment by yougoforward** [comment]
Yes, just convert fp16 input to fp32 at the beginning of the unsupported op and then convert the fp32 output to fp16 at the end of the op

**Comment by JingyuQian** [comment]
@yougoforward Thanks for the help. I have implemented this simple version, but unfortunately I didn't see any speed-up as expected. I might need to try some more complicated version.
